File: .claude/skills/flutter-workflow/scripts/create_workflow.py
#!/usr/bin/env python3
"""
Create GitHub Actions workflow for Flutter app in Melos workspace.

Usage:
    python create_workflow.py <app_path> [--name <workflow_name>]

Examples:
    python create_workflow.py apps/my_app
    python create_workflow.py apps/my_app --name ci.yml
"""

import argparse
import os
import re
import shutil
import sys
from pathlib import Path


def find_workspace_root() -> Path:
    """Find the workspace root by looking for pubspec.yaml with melos config."""
    current = Path.cwd()
    while current != current.parent:
        pubspec = current / "pubspec.yaml"
        if pubspec.exists():
            content = pubspec.read_text()
            if "melos:" in content or "workspace:" in content:
                return current
        current = current.parent
    return Path.cwd()


def create_workflow(app_path: str, workflow_name: str = "main.yml", workspace_root: Path = None):
    """Create workflow file from template."""
    if workspace_root is None:
        workspace_root = find_workspace_root()
    
    # Normalize app path
    app_path = app_path.rstrip("/")
    if not app_path.startswith("apps/") and not app_path.startswith("packages/"):
        # Assume it's under apps/
        if "/" not in app_path:
            app_path = f"apps/{app_path}"
    
    # Check if app exists
    full_app_path = workspace_root / app_path
    if not full_app_path.exists():
        print(f"Error: App path does not exist: {full_app_path}")
        sys.exit(1)
    
    # Locate template
    script_dir = Path(__file__).parent
    template_path = script_dir.parent / "assets" / "main.yml.template"
    
    if not template_path.exists():
        print(f"Error: Template not found: {template_path}")
        sys.exit(1)
    
    # Read template
    content = template_path.read_text()
    
    # Replace placeholder with app path
    content = content.replace("apps/__APP_NAME__", app_path)
    
    # Create .github/workflows directory
    workflows_dir = workspace_root / ".github" / "workflows"
    workflows_dir.mkdir(parents=True, exist_ok=True)
    
    # Write workflow file
    output_path = workflows_dir / workflow_name
    output_path.write_text(content)
    
    print(f"✅ Created workflow: {output_path}")
    print(f"   App path: {app_path}")
    return output_path


def main():
    parser = argparse.ArgumentParser(
        description="Create GitHub Actions workflow for Flutter app"
    )
    parser.add_argument("app_path", help="Path to the app (e.g., apps/my_app)")
    parser.add_argument("--name", default="main.yml", 
                       help="Workflow filename (default: main.yml)")
    parser.add_argument("--workspace", type=str,
                       help="Workspace root path (auto-detected if not specified)")
    
    args = parser.parse_args()
    
    workspace_root = Path(args.workspace) if args.workspace else None
    create_workflow(args.app_path, args.name, workspace_root)


if __name__ == "__main__":
    main()

File: .claude/skills/flutter-workflow/scripts/create_workflow.dart
#!/usr/bin/env dart
/// 为 Melos 工作区中的 Flutter 应用创建 GitHub Actions 工作流。
///
/// 用法:
///     dart run create_workflow.dart <应用路径> [--name <工作流名称>]
///
/// 示例:
///     dart run create_workflow.dart apps/my_app
///     dart run create_workflow.dart apps/my_app --name ci.yml

import 'dart:io';

import 'package:args/args.dart';
import 'package:path/path.dart' as path;
import 'package:project_workspace/project_workspace.dart';

/// 从模板创建工作流文件。
File createWorkflow(
  String appPath, {
  String workflowName = 'main',
  required Directory workspaceRoot,
  required Directory scriptDir,
}) {
  // 规范化应用路径
  appPath = appPath.replaceAll(RegExp(r'/+$'), '');
  if (!appPath.startsWith('apps/') && !appPath.startsWith('packages/')) {
    // 假设在 apps/ 下
    if (!appPath.contains('/')) {
      appPath = 'apps/$appPath';
    }
  }

  // 检查应用是否存在
  final fullAppPath = Directory(path.join(workspaceRoot.path, appPath));
  if (!fullAppPath.existsSync()) {
    logger.e('应用路径不存在: ${fullAppPath.path}');
    exit(1);
  }

  // 定位模板
  final templatePath = File(
    path.join(scriptDir.parent.path, 'assets', 'main.yml.template'),
  );

  if (!templatePath.existsSync()) {
    logger.e('模板未找到: ${templatePath.path}');
    exit(1);
  }

  // 读取模板
  var content = templatePath.readAsStringSync();

  // 替换占位符为应用路径
  content = content.replaceAll('apps/__APP_NAME__', appPath);

  // 创建 .github/workflows 目录
  final workflowsDir = Directory(
    path.join(workspaceRoot.path, '.github', 'workflows'),
  );
  if (!workflowsDir.existsSync()) {
    workflowsDir.createSync(recursive: true);
  }

  // 确保工作流名称有 .yml 扩展名
  if (!workflowName.endsWith('.yml') && !workflowName.endsWith('.yaml')) {
    workflowName = '$workflowName.yml';
  }

  // 写入工作流文件
  final outputPath = File(path.join(workflowsDir.path, workflowName));
  outputPath.writeAsStringSync(content);

  logger.i('✅ 已创建工作流: ${outputPath.path}');
  logger.i('   应用路径: $appPath');
  return outputPath;
}

void main(List<String> arguments) {

  final parser = ArgParser()
    ..addOption(
      'name',
      abbr: 'n',
      help: '工作流文件名（默认: main）',
      defaultsTo: 'main',
    )
    ..addOption(
      'workspace',
      abbr: 'w',
      help: '工作区根目录路径（未指定则自动检测）',
    )
    ..addFlag(
      'help',
      abbr: 'h',
      help: '显示帮助信息',
      negatable: false,
    );

  ArgResults args;
  List<String> rest;
  try {
    args = parser.parse(arguments);
    rest = args.rest;
  } catch (e) {
    print('错误: $e\n');
    print('用法: dart run create_workflow.dart <应用路径> [选项]');
    print('');
    print(parser.usage);
    exit(1);
  }

  if (args['help'] as bool || rest.isEmpty) {
    print('为 Flutter 应用创建 GitHub Actions 工作流');
    print('');
    print('用法: dart run create_workflow.dart <应用路径> [选项]');
    print('');
    print('示例:');
    print('  dart run create_workflow.dart apps/my_app');
    print('  dart run create_workflow.dart apps/my_app --name ci');
    print('');
    print('选项:');
    print(parser.usage);
    exit(args['help'] as bool ? 0 : 1);
  }

  final appPath = rest.first;
  final workflowName = args['name'] as String;
  final scriptPath = Platform.script.toFilePath();
  final scriptDir = Directory(path.dirname(scriptPath));
  final workspaceRoot = resolveWorkspace(args, scriptPath);

  createWorkflow(
    appPath,
    workflowName: workflowName,
    workspaceRoot: workspaceRoot,
    scriptDir: scriptDir,
  );
}
